### PR TITLE
Refactor ComposerPhpVersionFactory, ConstantResolver

### PR DESCRIPTION
**Pull request by staabm**
as requested in https://github.com/phpstan/phpstan-src/pull/3609#pullrequestreview-2428906861

### CODE REVIEW

File: src/Analyser/ConstantResolver.php [hunk 1]
```diff
@@ -35,12 +39,13 @@ final class ConstantResolver
...
 	 * @param string[] $dynamicConstantNames
+	 * @param int|array{min: int, max: int}|null $phpVersion
 	 */
 	public function __construct(
 		private ReflectionProviderProvider $reflectionProviderProvider,
 		private array $dynamicConstantNames,
-		private ?PhpVersion $composerMinPhpVersion,
-		private ?PhpVersion $composerMaxPhpVersion,
+		private int|array|null $phpVersion,
+		private ?ComposerPhpVersionFactory $composerPhpVersionFactory,
```
Comment (ondrejmirtes) [inline src/Analyser/ConstantResolver.php]: Why is this nullable?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/ConstantResolver.php (hunk 1 above)
Comment (staabm) [reply src/Analyser/ConstantResolver.php]: to allow [creation of the ConstantResolver](https://github.com/phpstan/phpstan-src/blob/38630336ab201cb9ecce6d4104ec1ea5b08e8b7e/src/DependencyInjection/ValidateIgnoredErrorsExtension.php#L68) when the DI is not yet constructed

Comment (ondrejmirtes) [reply src/Analyser/ConstantResolver.php]: Just pass `new ComposerPhpVersionFactory` with empty `$composerAutoloaderProjectPaths` there.

File: src/Analyser/ConstantResolver.php [hunk 2]
```diff
@@ -116,25 +121,25 @@ public function resolvePredefinedConstant(string $resolvedConstantName): ?Type
...
-				&& $this->composerMaxPhpVersion->getMajorVersionId() === $this->composerMinPhpVersion->getMajorVersionId()
-				&& $this->composerMaxPhpVersion->getMinorVersionId() === $this->composerMinPhpVersion->getMinorVersionId()
+				$this->getMinPhpVersion() !== null
+				&& $this->getMaxPhpVersion() !== null
+				&& $this->getMaxPhpVersion()->getMajorVersionId() === $this->getMinPhpVersion()->getMajorVersionId()
+				&& $this->getMaxPhpVersion()->getMinorVersionId() === $this->getMinPhpVersion()->getMinorVersionId()
 			) {
-				$minRelease = $this->composerMinPhpVersion->getPatchVersionId();
-				$maxRelease = $this->composerMaxPhpVersion->getPatchVersionId();
+				$minRelease = $this->getMinPhpVersion()->getPatchVersionId();
```
Comment (ondrejmirtes) [inline src/Analyser/ConstantResolver.php]: You're calling the getter multiple times and it always leads to creating new object. I'd just do `if (in_array($resolvedConstantName, ['PHP_VERSION_ID', 'PHP_MAJOR_VERSION', 'PHP_MINOR_VERSION']))` and save it to a variable first, then use the variable. inside for each constant.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Thank you! Now onto the diagnose command :)

**Pull request merged**